polyglot/piranha/src/models/source_code_unit.rs

```diff
@@ -79,55 +80,110 @@ impl SourceCodeUnit {
+  pub(crate) fn _replace_file_contents_and_re_parse(
+    &mut self, replacement_content: &str, parser: &mut Parser, is_current_ast_edited: bool,
+  ) {
+    let prev_tree = if is_current_ast_edited {
+      Some(&self.ast)
+    } else {
+      None
+    };
+    println!("Replacing file contents");
     // Create a new updated tree from the previous tree
     let new_tree = parser
-      .parse(&new_source_code, Some(&self.ast))
+      .parse(&replacement_content, prev_tree)
       .expect("Could not generate new tree!");
     self.ast = new_tree;
-    self.code = new_source_code;
-    // Handle errors, like removing extra comma.
-    self.remove_additional_comma_from_sequence_list(parser);
-    if self.ast.root_node().has_error() {
-      panic!(
-        "Produced syntactically incorrect source code {}",
-        self.code()
-      );
+    self.code = replacement_content.to_string();
+  }
+
+  // Tries to remove the extra comma -
+  // -->  Remove comma if extra
+  //    --> Check if AST has no errors, to decide if the replacement was successful. 
+  //      --->  if No: Undo the change
+  //      --->  if Yes: Return   
+  // Returns true if the comma was successfully removed.
+  fn try_to_remove_extra_comma(&mut self, parser: &mut Parser) -> bool {
+    let c_ast = self.ast.clone();
+    for n in traverse(c_ast.walk(), Order::Post) {
+      // Remove the extra comma
+      if n.is_extra() && eq_without_whitespace(n.utf8_text(self.code().as_bytes()).unwrap(), ",") {
+        let current_version_code = self.code().clone();
+        self._apply_edit(n.range(), "", parser, false);
+        if self.ast.root_node().has_error() {
+          // Undo the edit applied above
+          self._replace_file_contents_and_re_parse(&current_version_code, parser, false);
+        } else {
+          return true;
+        }
+      }
     }
-    ts_edit
+    false
+  }
+
+  // Tries to remove the extra comma - 
+  // Applies three Regex-Replacements strategies to the source file to remove the extra comma. 
+  // *  replace consecutive commas with comma 
+  // *  replace ( `(,` or `[,` ) with just `(` or `[`)
+  // Check if AST has no errors, to decide if the replacement was successful. 
+  // Returns true if the comma was successfully removed.
+  fn try_to_fix_code_with_regex_replace(&mut self, parser: &mut Parser) -> bool{
+    let consecutive_comma_pattern = Regex::new(r",\s*\n*,").unwrap();
+    let square_bracket_comma_pattern = Regex::new(r"\[\s*\n*,").unwrap();
+    let round_bracket_comma_pattern = Regex::new(r"\(\s*\n*,").unwrap();
+    let strategies = [(consecutive_comma_pattern, ","), (square_bracket_comma_pattern, "["), (round_bracket_comma_pattern, "(")];
+
+    let mut content = self.code();
+    for (regex_pattern, replacement ) in strategies{
+        if regex_pattern.is_match(&content){
+          content = regex_pattern.replace_all(&content, replacement).to_string();
+          self._replace_file_contents_and_re_parse(&content, parser, false);
```

Comment:
[Reference](https://tree-sitter.github.io/tree-sitter/#underlying-research) - Under the hood tree-sitter uses incremental parsing algorithm . So it is not expensive to modify and parse again, since re-parsing is done incrementally (i.e. it it will be done at statement level not compilation unit level)